File: client/src/main/scala/stasis/client/api/clients/Clients.scala
package stasis.client.api.clients

import stasis.core.discovery.providers.client.ServiceDiscoveryProvider

sealed trait Clients {
  def api: ServerApiEndpointClient
  def core: ServerCoreEndpointClient
}

object Clients {
  final case class Static(
    override val api: ServerApiEndpointClient,
    override val core: ServerCoreEndpointClient
  ) extends Clients

  final case class Discovered(
    discovery: ServiceDiscoveryProvider
  ) extends Clients {
    override def api: ServerApiEndpointClient = discovery.latest[ServerApiEndpointClient]
    override def core: ServerCoreEndpointClient = discovery.latest[ServerCoreEndpointClient]
  }

  def apply(api: ServerApiEndpointClient, core: ServerCoreEndpointClient): Clients =
    Clients.Static(api = api, core = core)

  def apply(discovery: ServiceDiscoveryProvider): Clients =
    Clients.Discovered(discovery = discovery)
}

File: client-android/lib/src/main/kotlin/stasis/client_android/lib/api/clients/Clients.kt
package stasis.client_android.lib.api.clients

import stasis.client_android.lib.discovery.exceptions.DiscoveryFailure
import stasis.client_android.lib.discovery.providers.client.ServiceDiscoveryProvider
import stasis.client_android.lib.utils.Try
import stasis.client_android.lib.utils.Try.Failure
import java.util.concurrent.atomic.AtomicReference

sealed interface Clients {
    val api: ServerApiEndpointClient
    val core: ServerCoreEndpointClient

    fun withDiscovery(discovery: Try<ServiceDiscoveryProvider>)

    data class Static(
        override val api: ServerApiEndpointClient,
        override val core: ServerCoreEndpointClient
    ) : Clients {
        override fun withDiscovery(discovery: Try<ServiceDiscoveryProvider>) = Unit // do nothing
    }

    class Discovered : Clients {
        private val discoveryRef: AtomicReference<Try<ServiceDiscoveryProvider>> =
            AtomicReference(Failure(DiscoveryFailure("No discovery provider found")))

        override val api: ServerApiEndpointClient
            get() = discoveryRef.get().get().latest(ServerApiEndpointClient::class.java)

        override val core: ServerCoreEndpointClient
            get() = discoveryRef.get().get().latest(ServerCoreEndpointClient::class.java)

        override fun withDiscovery(discovery: Try<ServiceDiscoveryProvider>) =
            discoveryRef.set(discovery)
    }

    companion object {
        operator fun invoke(api: ServerApiEndpointClient, core: ServerCoreEndpointClient): Clients =
            Static(api = api, core = core)

        fun discovered(): Clients =
            Discovered()
    }
}
